File: benchmarks/julia/benchmark_matrix_lu.jl
#!/usr/bin/env julia

import Pkg

Pkg.activate(; temp=true)
if haskey(ENV, "ACI_JL_PATH")
    Pkg.develop(path=ENV["ACI_JL_PATH"])
else
    Pkg.add(url="https://github.com/tensor4all/AlternatingCrossInterpolation.jl.git")
end

using Statistics
import AlternatingCrossInterpolation as ACI

const TCI = ACI.TCI

struct MatrixLuTiming
    inplace_ns::Int
    borrowed_ns::Int
    rank::Int
    last_error::Float64
    checksum::Float64
end

function parse_sizes(args)
    sizes = [16, 32, 64, 128]
    i = 1
    while i <= length(args)
        if args[i] == "--sizes"
            i == length(args) && error("--sizes requires a comma-separated value list")
            sizes = parse.(Int, split(args[i + 1], ","))
            i += 2
        else
            error("unknown argument: $(args[i])")
        end
    end
    return sizes
end

function hilbert_matrix(size::Int)
    A = Matrix{Float64}(undef, size, size)
    for col in 1:size, row in 1:size
        A[row, col] = 1.0 / (row + col - 1)
    end
    return A
end

function timed_once(size::Int, leftorthogonal::Bool)
    A = hilbert_matrix(size)

    A_copy = copy(A)
    t = time_ns()
    lu_inplace = TCI.rrlu!(
        A_copy;
        leftorthogonal=leftorthogonal,
        maxrank=typemax(Int),
        reltol=0.0,
        abstol=1.0e-10,
    )
    inplace_ns = time_ns() - t

    t = time_ns()
    lu_borrowed = TCI.rrlu(
        A;
        leftorthogonal=leftorthogonal,
        maxrank=typemax(Int),
        reltol=0.0,
        abstol=1.0e-10,
    )
    borrowed_ns = time_ns() - t

    left = TCI.left(lu_inplace; permute=false)
    right = TCI.right(lu_inplace; permute=false)
    checksum = (isempty(left) ? 0.0 : abs(left[1, 1])) +
        (isempty(right) ? 0.0 : abs(right[1, 1]))

    return MatrixLuTiming(
        inplace_ns,
        borrowed_ns,
        TCI.npivots(lu_borrowed),
        TCI.lastpivoterror(lu_borrowed),
        checksum,
    )
end

ns_to_ms(ns::Int) = ns / 1e6

function main()
    repeats = parse(Int, get(ENV, "T4A_MATRIX_LU_REPEATS", "20"))
    sizes = parse_sizes(ARGS)
    println("impl,matrix,size,repeats,left_orthogonal,inplace_ms,borrowed_ms,rank,last_error,checksum")
    for size in sizes
        for leftorthogonal in (true, false)
            runs = [timed_once(size, leftorthogonal) for _ in 1:repeats]
            first = runs[1]
            println(join([
                "julia",
                "hilbert",
                size,
                repeats,
                leftorthogonal,
                median(ns_to_ms.(getfield.(runs, :inplace_ns))),
                median(ns_to_ms.(getfield.(runs, :borrowed_ns))),
                first.rank,
                first.last_error,
                first.checksum,
            ], ","))
        end
    end
end

main()

File: benchmarks/rust/benchmark_matrix_lu.rs
use std::time::{Duration, Instant};

use tensor4all_tcicore::{rrlu, rrlu_inplace, RrLUOptions};
use tensor4all_tensorbackend::Matrix;

#[derive(Clone, Copy, Default)]
struct MatrixLuTiming {
    inplace: Duration,
    borrowed: Duration,
    rank: usize,
    last_error: f64,
    checksum: f64,
}

fn hilbert_matrix(size: usize) -> Matrix<f64> {
    let mut data = vec![0.0; size * size];
    for col in 0..size {
        for row in 0..size {
            data[row + size * col] = 1.0 / ((row + col + 1) as f64);
        }
    }
    Matrix::from_col_major_vec(size, size, data)
}

fn parse_env_usize(name: &str, default: usize) -> usize {
    std::env::var(name)
        .ok()
        .and_then(|value| value.parse::<usize>().ok())
        .unwrap_or(default)
}

fn parse_env_usize_list(name: &str, default: Vec<usize>) -> Vec<usize> {
    std::env::var(name)
        .ok()
        .map(|value| {
            value
                .split(',')
                .map(|item| item.parse::<usize>().unwrap())
                .collect::<Vec<_>>()
        })
        .unwrap_or(default)
}

fn timing_ms(duration: Duration) -> f64 {
    duration.as_secs_f64() * 1_000.0
}

fn timing_median(mut values: Vec<f64>) -> f64 {
    values.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let mid = values.len() / 2;
    if values.len().is_multiple_of(2) {
        0.5 * (values[mid - 1] + values[mid])
    } else {
        values[mid]
    }
}

fn timed_hilbert_matrix_lu_once(size: usize, left_orthogonal: bool) -> MatrixLuTiming {
    let matrix = hilbert_matrix(size);
    let options = RrLUOptions {
        max_rank: usize::MAX,
        rel_tol: 0.0,
        abs_tol: 1.0e-10,
        left_orthogonal,
    };

    let mut matrix_for_inplace = matrix.clone();
    let start = Instant::now();
    let lu_inplace = rrlu_inplace(&mut matrix_for_inplace, Some(options.clone())).unwrap();
    let inplace = start.elapsed();

    let start = Instant::now();
    let lu_borrowed = rrlu(&matrix, Some(options)).unwrap();
    let borrowed = start.elapsed();

    let left = lu_inplace.left(false);
    let right = lu_inplace.right(false);
    let checksum = if left.nrows() > 0 && left.ncols() > 0 {
        left[[0, 0]].abs()
    } else {
        0.0
    } + if right.nrows() > 0 && right.ncols() > 0 {
        right[[0, 0]].abs()
    } else {
        0.0
    };

    MatrixLuTiming {
        inplace,
        borrowed,
        rank: lu_borrowed.npivots(),
        last_error: lu_borrowed.last_pivot_error(),
        checksum,
    }
}

fn main() {
    let repeats = parse_env_usize("T4A_MATRIX_LU_REPEATS", 20);
    let sizes = parse_env_usize_list("T4A_MATRIX_LU_SIZES", vec![16, 32, 64, 128]);

    println!(
        "impl,matrix,size,repeats,left_orthogonal,inplace_ms,borrowed_ms,rank,last_error,checksum"
    );
    for size in sizes {
        for left_orthogonal in [true, false] {
            let runs = (0..repeats)
                .map(|_| timed_hilbert_matrix_lu_once(size, left_orthogonal))
                .collect::<Vec<_>>();
            let first = runs[0];
            println!(
                "rust,hilbert,{size},{repeats},{left_orthogonal},{:.6},{:.6},{},{:.6e},{:.6e}",
                timing_median(runs.iter().map(|run| timing_ms(run.inplace)).collect()),
                timing_median(runs.iter().map(|run| timing_ms(run.borrowed)).collect()),
                first.rank,
                first.last_error,
                first.checksum,
            );
        }
    }
}
